content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+---
+title: "Random numbers done right"
+date: 2022-04-25
+draft: false
+description: "
+Use `np.random.Generator` and don't seed! A quick explainer on random number
+generation with NumPy.
+"
+tags: ["NumPy", "tutorials", "random-numbers"]
+displayInList: true
+author: ["Pamphile Roy"]
+
+---
+
+Do you need random numbers? You use NumPy? NumPy has a canonical way to
+produce random numbers with
+[`np.random.Generator`](https://numpy.org/doc/stable/reference/random/index.html).
+There is a say in Python, "global variables are bad". And you guessed correctly,
+it's also bad to use `np.random.random()` or other generators that rely on a
+global state! I will also demonstrate why you should refrain yourself from
+using a seed.
+
+In the following, we assume that _NumPy_, _SciPy_ and _Matplotlib_ are
+installed and imported:
+
+```python
+import numpy as np
+import scipy.stats as stats
+import matplotlib.pyplot as plt
+```
+
+## Random done wrong
+
+```python
+# the following is wrong and only exists for demonstration purposes!
+np.random.seed(123456)
+sample = np.random.random(1_024)
+```
+
+Aoutch! Two lines of codes and both are wrong. This is a simple snippet to
```

Comment:
Strong word, "wrong" :).  I don't mind it too much, I just tend to try to be a bit less strong.

Reply:
I think it's too strong and misdirected at the writers of such code. If anyone was wrong, it was us (half me for putting the global in the first place, half someone else who added `np.random.seed()` back in the mists of Subversion). People are just using the APIs we designed for them. It is, indeed, suboptimal code, IMO, but I don't think it's wrong. For many purposes, it's quite fine.